security/hcaptcha.go

```diff
@@ -0,0 +1,87 @@
+package security
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type CaptchaResponse struct {
+	Token string `json:"hcaptcha_token"`
+}
+
+type VerificationResponse struct {
+	Success    bool     `json:"success"`
+	ErrorCodes []string `json:"error-codes"`
+	Hostname   string   `json:"hostname"`
+}
+
+type VerificationResult int
+const (
+	UserRequestFailed VerificationResult = iota
+	VerificationProcessFailure
+	SuccessfullyVerified
+)
+var Client *http.Client
+
+func init() {
+	// TODO (darora): make timeout configurable
+	Client = &http.Client{Timeout: 10 * time.Second}
+}
+
+func VerifyRequest(r *http.Request, secretKey string) (VerificationResult, error) {
+	res := CaptchaResponse{}
+	bodyBytes, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return UserRequestFailed, err
+	}
+	r.Body.Close()
+	// re-init body so downstream route handlers don't get borked
+	r.Body = ioutil.NopCloser(bytes.NewBuffer(bodyBytes))
+
+	jsonDecoder := json.NewDecoder(bytes.NewBuffer(bodyBytes))
+	err = jsonDecoder.Decode(&res)
+	if err != nil || strings.TrimSpace(res.Token) == "" {
+		return UserRequestFailed, errors.Wrap(err, "couldn't decode captcha info")
+	}
+	clientIP := strings.Split(r.RemoteAddr, ":")[0]
+	return verifyCaptchaCode(res.Token, secretKey, clientIP)
+}
+
+func verifyCaptchaCode(token string, secretKey string, clientIP string) (VerificationResult, error) {
+	data := url.Values{}
+	data.Set("secret", secretKey)
+	data.Set("response", token)
+	data.Set("remoteip", clientIP)
+	// TODO (darora): pipe through sitekey
+
+	r, err := http.NewRequest("POST", "https://hcaptcha.com/siteverify", strings.NewReader(data.Encode()))
+	if err != nil {
+		return VerificationProcessFailure, errors.Wrap(err, "couldn't initialize request object for hcaptcha check")
+	}
+	r.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	r.Header.Add("Content-Length", strconv.Itoa(len(data.Encode())))
+	res, err := Client.Do(r)
+	if err != nil {
+		return VerificationProcessFailure, errors.Wrap(err, "failed to verify hcaptcha token")
+	}
+	verResult := VerificationResponse{}
+	decoder := json.NewDecoder(res.Body)
```

Comment:
necessary to `defer res.Body.Close()` here? seems like it's recommended in the go package [here](https://pkg.go.dev/net/http)

Reply:
done